how-to/customize-workspace/client/src/endpoint.ts

```diff
@@ -0,0 +1,99 @@
+import { getSettings } from "./settings";
+import { Endpoint } from "./shapes";
+
+let availableEndpoints: Endpoint<unknown>[];
+
+interface FetchOptions { body?:string, method?: "GET"|"POST", credentials?: "omit" | "same-origin" | "include", mode?: "no-cors" | "cors" | "same-origin", cache?:"default" | "no-cache" | "reload" | "force-cache" | "only-if-cached", redirect?: "manual" | "follow" | "error", referrerPolicy?:"no-referrer" | "no-referrer-when-downgrade" | "origin" | "origin-when-cross-origin" | "same-origin" | "strict-origin" | "strict-origin-when-cross-origin" | "unsafe-url", headers?: {[key:string]:string }  };
+
+export async function init(endpoints: Endpoint<unknown>[] = null) {
+  if(endpoints !== null) {
+    availableEndpoints = endpoints; 
+  } else {
+    let settings = await getSettings();
+    availableEndpoints = settings.endpointProvider.endpoints || [];
+  }
+}
+
+function getOptions(options:any): FetchOptions {
+  
+  const { url, ...fetchOptions } = options;
+  
+  return fetchOptions;
+}
+
+function getUrl(options:{url?:string}): string {
+  return options.url;
+}
+
+function getRequestOptions(url: string, options:FetchOptions, request) : { url:string, options: FetchOptions } {
+  if(options.method === "GET") {
+    if(request !== undefined) {
+      let keys = Object.keys(request);
+      if(keys.length > 0) {
+        let length = keys.length;
+        for(let i = 0; i < length; i++) {
+          url = url.replace(`[${keys[i]}]`,request[keys[i]] );
```

Comment:
The `request[keys[i]]` should probably be wrapped with encodeURIComponent

Reply:
yeah good call. Done